content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
++++
+title = "Compiler Validation via Equivalence Modulo Inputs"
+[extra]
+bio = """
+  [Rolph Recto](rolph-recto.github.io) is a third-year graduate student studying
+  the intersection of programming languages, security, and distributed systems.
+  [Gregory Yauney](https://www.cs.cornell.edu/~gyauney)
+  is a second-year student working on machine learning and
+  digital humanities.
+"""
+latex = true
+[[extra.authors]]
+name = "Rolph Recto"
+link = "rolph-recto.github.io"
+[[extra.authors]]
+name = "Gregory Yauney"
+link = "https://www.cs.cornell.edu/~gyauney"
++++
+
+
+Imagine you, a clever person who wants her C programs to run faster,
+sat down, thought very hard, and developed a new compiler optimization.
+Say you implement it as a transformation pass in LLVM so that other people
+can take advantage of your cleverness to make their programs run faster as well.
+You run your optimization pass over a few benchmarks and see that it does
+indeed make some programs run faster.
+But a question nags you: how do you know that your optimization is correct?
+That is, how do you know that your optimization doesn't change the semantics
+of the input program?
+
+*Equivalence Modulo Inputs* (henceforth EMI), a testing technique introduced
+by Le & al. in a [PLDI 2014 paper][paper], allows our compiler hacker
+above to test her optimization rigorously without much effort.
+EMI is especially effective at finding *miscompilation* bugs,
+wherein compilers produce wrong code, which is much more pernicious than
+*compiler crashes*, bugs where the compiler terminates abnormally.
+This allows EMI to test the optimization phase of compilers much more rigorously
+than prior work such as [Csmith][], which mostly finds compiler crashes and not
```

Comment:
Maybe *fewer* miscompilations? To avoid the inadvertent implication that Csmith is incapable of finding miscompilations.